backend/editor/OT/OTClient/operation.ts

```diff
@@ -1,165 +1,252 @@
-    a: number[],
-    b: number[],
-    tp: number
-  ): [number[], number[]] => {
-    switch (true) {
-      case a[tp] > b[tp]:
-        return [update(a, tp, -1), b];
-      case a[tp] < b[tp]:
-        return [a, update(b, tp, -1)];
-      default:
-        return a.length > b.length
-          ? [[], b]
-          : (a.length < b.length
-            ? [a, []]
-            : [[], []]);
+  }
+
+  return [a.path, b.path];
+};
+
+/**
+ * Takes 2 paths and their transformation point and transforms them as if they
+ * were insertion functions
+ */
+const transformInserts = (
+  a: number[],
+  b: number[],
+  tp: number
+): [number[], number[]] => {
+  switch (true) {
+    case a[tp] > b[tp]:
+      return [update(a, tp, 1), b];
+    case a[tp] < b[tp]:
+      return [a, update(b, tp, 1)];
+    default:
+      return a.length > b.length
+        ? [update(a, tp, 1), b]
+        : a.length < b.length
+        ? [a, update(b, tp, 1)]
+        : [a, b];
+  }
+};
+
+/**
+ * Takes 2 paths and transforms them as if they were deletion operations
+ */
+const transformDeletes = (
+  a: number[],
+  b: number[],
+  tp: number
+): [number[], number[]] => {
+  switch (true) {
+    case a[tp] > b[tp]:
+      return [update(a, tp, -1), b];
+    case a[tp] < b[tp]:
+      return [a, update(b, tp, -1)];
+    default:
+      return a.length > b.length
+        ? [[], b]
+        : a.length < b.length
+        ? [a, []]
+        : [[], []];
+  }
+};
+
+/**
+ * Takes an insertion operation and a deletion operation and transforms them
+ */
+const transformInsertDelete = (
+  insertOp: number[],
+  deleteOp: number[],
+  tp: number
+): [number[], number[]] => {
+  switch (true) {
+    case insertOp[tp] > deleteOp[tp]:
+      return [update(insertOp, tp, -1), deleteOp];
+    case insertOp[tp] < deleteOp[tp]:
+      return [insertOp, update(deleteOp, tp, 1)];
+    default:
+      return insertOp.length > deleteOp.length
+        ? [[], deleteOp]
+        : [insertOp, update(deleteOp, tp, 1)];
+  }
+};
+
+/**
+ * Updates a specific index in a path
+ */
+const update = (pos: number[], toChange: number, change: number) => {
+  pos[toChange] += change;
+  return pos;
+};
+
+/**
+ * Takes in two paths and computes their transformation point
+ */
+const transformationPoint = (a: number[], b: number[]): number =>
+  [...Array(Math.min(a.length, b.length)).keys()].find((i) => a[i] != b[i]) ??
+  Math.min(a.length, b.length);
+
+/**
+ * Takes two paths and determines if their effect is independent or not
+ */
+const effectIndependent = (a: number[], b: number[], tp: number): boolean =>
+  (a.length > tp + 1 && b.length > tp + 1) ||
+  (a[tp] > b[tp] && a.length < b.length) ||
+  (a[tp] < b[tp] && a.length > b.length);
+
+/**
+ * Normalise turns an empty operation into a noop
+ */
+const normalise = (a: Operation): Operation => (a.path.length === 0 ? noOp : a);
+
+const copy = <T>(a: T): T => JSON.parse(JSON.stringify(a));
+
+export const stringTransform = (
```

Comment:
So the general idea behind these type specific transformation functions is that we invoke these transformation functions whenever our operations don't morph the document tree but could still result in conflicts (at the level of individual properties), I think it might be worthwhile doing a simmilar thing to what we do in Go where we define an EditType interface which contains this Transform funtion and then invoke that whenever we run into the above situation

Reply:
I.e we have something like:
```ts
interface AtomicOperation {
   $type: string;
   transformAgainst: (op: AtomicOperation) => {AtomicOperation, AtomicOperation}
   // We leave out apply as our client is only concerned with transformation not application
}

class StringOperation implements AtomicOperation {
   $type: string = "stringOperation";
   transformAgainst = (op: AtomicOperation) => {
    // your string transformation code
   }
}
```
So then within our actual general transformation functions once we detect a scenario where the conflict is at the level of invidual values and not the document tree we can invoke:
```ts
Operation.operation.transformAgainst(...)
```
On a sidenote, these properties should probably have better names :P `Operation.operation` is rather odd

Reply:
nah it's good to me